src/Microsoft.OData.Edm/Csdl/Serialization/EdmModelCsdlSchemaJsonWriter.cs

```diff
@@ -0,0 +1,1292 @@
+            this.jsonWriter.WriteStartObject();
+
+            // The entity container object MUST contain the member $Kind with a string value of EntityContainer.
+            // Be caution: Example 33 has the $Kind, but no other words.
+            this.jsonWriter.WriteRequiredProperty("$Kind", CsdlConstants.Element_EntityContainer);
+
+            // The entity container object MAY contain the member $Extends, so far it only supports in Csdl Semantics.
+            CsdlSemanticsEntityContainer tmp = container as CsdlSemanticsEntityContainer;
+            CsdlEntityContainer csdlContainer = null;
+            if (tmp != null && (csdlContainer = tmp.Element as CsdlEntityContainer) != null)
+            {
+                this.jsonWriter.WriteOptionalProperty("$Extends", csdlContainer.Extends);
+            }
+        }
+
+        /// <summary>
+        /// Write Entity Set object
+        /// </summary>
+        /// <param name="entitySet">The Edm entity set.</param>
+        internal override void WriteEntitySetElementHeader(IEdmEntitySet entitySet)
+        {
+            // An entity set is represented as a member of the entity container object whose name is the name of the entity set
+            this.jsonWriter.WritePropertyName(entitySet.Name);
+
+            // whose value is an object
+            this.jsonWriter.WriteStartObject();
+
+            // The entity set object MUST contain the members $Collection, it's value as true
+            this.jsonWriter.WriteRequiredProperty("$Collection", true);
+
+            // The entity set object MUST contain the member $Type whose string value is the qualified name of an entity type.
+            this.jsonWriter.WriteRequiredProperty("$Type", entitySet.EntityType().FullName());
+
+            // It MAY contain the members $IncludeInServiceDocument. Absence of the member means true.
+            this.jsonWriter.WriteOptionalProperty("$IncludeInServiceDocument", entitySet.IncludeInServiceDocument, true);
+        }
+
+        /// <summary>
+        /// Write Singleton object
+        /// </summary>
+        /// <param name="singleton">The Edm singleton.</param>
+        internal override void WriteSingletonElementHeader(IEdmSingleton singleton)
+        {
+            // A singleton is represented as a member of the entity container object whose name is the name of the singleton
+            this.jsonWriter.WritePropertyName(singleton.Name);
+
+            // whose value is an object
+            this.jsonWriter.WriteStartObject();
+
+            // The singleton object MUST contain the member $Type whose string value is the qualified name of an entity type.
+            this.jsonWriter.WriteRequiredProperty("$Type", singleton.EntityType().FullName());
+
+            // The singleton object MAY contain the member $Nullable. In OData 4.0 responses this member MUST NOT be specified.
+            // So far, IEdmSingleton doesn't have the property defined, so skip it now.

```

Comment:
>// So far, IEdmSingleton doesn't have the property defined, so skip it now. [](start = 12, length = 75)

Should probably open an issue for this.  What happens on read?

Reply:
Issues created at: [https://github.com/OData/odata.net/issues/1814](https://github.com/OData/odata.net/issues/1814)

at read side, I read this nullable value if this value is present. But, there's no way to use, so, just read it out.

See line 369 - 373 at SchemaJsonParser.cs

---
In reply to: [443753720](https://github.com/OData/odata.net/pull/1775#discussion_r443753720) [](ancestors = 443753720)